sw/device/tests/sram_ctrl_subword_access_test.c

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_sram_ctrl.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+#define SRAM_CTRL_TEST_DATA_SIZE_WORDS 3
+#define SRAM_CTRL_TEST_DATA_SIZE_BYTES (SRAM_CTRL_TEST_DATA_SIZE_WORDS * 4)
+
+OTTF_DEFINE_TEST_CONFIG();
+
+static const uint32_t kStatusRegMask = kDifSramCtrlStatusBusIntegErr |
+                                       kDifSramCtrlStatusInitErr |
+                                       kDifSramCtrlStatusEscalated;
+
+static const uint32_t kRandomData[SRAM_CTRL_TEST_DATA_SIZE_WORDS] = {
+    0x6b4abfae, 0x63bdb6e7, 0x87f99b1a};
+
+// Buffer to allow the compiler to allocate a safe area in Main SRAM where
+// we can do the write/read test without the risk of clobbering data
+// used by the program.
+OT_SECTION(".data")
+static volatile uint32_t sram_main_buffer[SRAM_CTRL_TEST_DATA_SIZE_WORDS];
+
+/// Write `kRandomData` to SRAM, then read back a u8, u16, u32, and u64 from
+/// each possible byte offset.
+static void read_subwords_check(mmio_region_t region) {
+  mmio_region_memcpy_to_mmio32(region, 0, kRandomData,
+                               SRAM_CTRL_TEST_DATA_SIZE_BYTES);
+
+  // Check byte reads.
+  for (uint32_t i = 0; i < SRAM_CTRL_TEST_DATA_SIZE_BYTES; ++i) {
+    uint8_t expected = ((volatile uint8_t *)kRandomData)[i];
+    uint8_t got = ((volatile uint8_t *)region.base)[i];
+    CHECK(expected == got,
+          "byte %d read back incorrectly: expected %02x, got %02x", i, expected,
+          got);
+  }
+
+  // Check uint16_t reads.
+  for (uint32_t i = 0; i < SRAM_CTRL_TEST_DATA_SIZE_BYTES - 1; ++i) {
+    uint16_t expected = *(volatile uint16_t *)((uint8_t *)kRandomData + i);
+    uint16_t got = *(volatile uint16_t *)((uint8_t *)region.base + i);
+    CHECK(expected == got,
+          "uint16_t %d read back incorrectly: expected %04x, got %04x", i,
+          expected, got);
+  }
+
+  // Check uint32_t reads.
+  for (uint32_t i = 0; i < SRAM_CTRL_TEST_DATA_SIZE_BYTES - 3; ++i) {
+    uint32_t expected = *(volatile uint32_t *)((uint8_t *)kRandomData + i);
+    uint32_t got = *(volatile uint32_t *)((uint8_t *)region.base + i);
+    CHECK(expected == got,
+          "uint32_t %d read back incorrectly: expected %08x, got %08x", i,
+          expected, got);
+  }
+
+  // Check uint64_t reads.
+  for (uint64_t i = 0; i < SRAM_CTRL_TEST_DATA_SIZE_BYTES - 7; ++i) {
+    uint64_t expected = *(volatile uint64_t *)((uint8_t *)kRandomData + i);
+    uint64_t got = *(volatile uint64_t *)((uint8_t *)region.base + i);
+    CHECK(expected == got,
+          "uint64_t %d read back incorrectly: expected %016x, got %016x", i,
+          expected, got);
```

Comment:
I think that OT print lib does not support uint64_t, so you need to use the trick below:
```suggestion
          "uint64_t %d read back incorrectly: expected %x%x, got %x%x", i,
          (uint32_t)expected>>32, (uint32_t)expected, (uint32_t)got >> 32, (uint32_t)got);
```